README.md
Austrian aid scraper
==============================
The scraper extracts information from the austrian development projects since 2010 from the austrian development agency website. The automatically extracted informations are stored in CSV and JSON files to make the further usage as easy as possible.

This repository provides the code and documentation and [keeps track of bugs as well as feature requests](https://github.com/OKFNat/aidScraper/issues).

- Team: [Gute Taten für gute Daten](http://okfn.at/gutedaten/) project (Open Knowledge Austria)
- Status: Production
- Documentation: English
- Licenses: [MIT License](http://opensource.org/licenses/MIT) + [Creative Commons Attribution 4.0](http://creativecommons.org/licenses/by/4.0/)

**Used software**
The sourcecode is written in Python 2. It was created with use of [iPython](http://ipython.org/), [BeautifulSoup4](http://www.crummy.com/software/BeautifulSoup/) and [urllib2](https://docs.python.org/2/library/urllib2.html).


## SCRAPER

**Description**

The scraper fetches the html passed in as urls from a csv file and stores them locally. The html is then parsed with BeautifulSoup4. Every table between the requested start country and end country is parsed out row by row, cell by cell and stored into a JSON structure with importing countries -> exporting countries -> arms classes -> data. The data structure is then  used to create nodes and edges files as JSON and CSV. This can also be used to extract country specific data to understand imports and exports from a country's perspective.

**Run scraper**
Go into the root folder of this repository and execute following commands in your terminal:
```
cd code
python aid-scraper.py
```

**Original sourcecode**

Thanks to Christian Goebel for the [original sourcecode](https://github.com/ChristianGoebel/Scrape_ADC), which got used for the final version.

### How the scraper works
**Configure the Scraper**

There are two global variables in [aid-scraper.py](code/aid-scraper.py) you may want to change to your needs.
- DELAY_TIME: To not overload the server or may get blocked because of too many request, you should set the delay time to fetch to 1-5 seconds, not less.
- TS: The timestamp as a string can be set to the last download. So you can use downloaded data over and over again and must not do it everytime. When you do it first time, you can set the value to ```datetime.now().strftime('%Y-%m-%d-%H-%M')```, so it is the timestamp when the scraper starts.

**Download the raw html data**

Here all the html raw data gets downloaded, stored locally and the basic data gets parsed.
- Download all overview pages with the tables (html). The navigation for the fetching runs through all overview pages by asking the existance of the "weiter" anchor and counting up an url variable.
- Open the downloaded files.
- Parse out the basic information about each project from the overview tables. This is necessary here, because the download of the project page needs the link from the overview table.
- Store the parsed data as JSON file.
- Download all project pages (html).

**Parse the html**

Here the description of the project gets added to the data.
- Open the JSON data.
- Open the project-pages files (html).
- Parse out the additional description information from the project pages.
- Store updated data as JSON file.

**Export as CSV**

Here the data gets exported as a CSV file.
- Open the data (JSON).
- Save the serialized data as CSV file.

## DATA INPUT
The original data is from the project list of the austrian development agency (ADA) [published on their website](http://www.entwicklung.at/zahlen-daten-und-fakten/projektliste/). The data consists of all contracts approved since January 1st of 2010.  in the list in chronologically descending order. The date of the last update can be found on the first table page as "Datum der letzten Aktualisierung".

### The Tables
The tables are the basic data, where most of the data is parsed out. The data is published in the following structure (e. g. first project).


| Vertragsnummer | Vertragstitel | Land/Region | OEZA/ADA-Vertragssumme | Vertragspartner |
|----------------|---------------|-------------|------------------------|-----------------|
| 2325-02/2016 | Programm zum Schutz der MenschenrechtsverteidigerInnen in der westlichen Region Guatemalas | Guatemala | EUR 64.300,00 | HORIZONT3000 - Österreichische Organisation für Entwicklungszusammena |

**Attributes**
- Vertragsnummer: contract number of project.
- Vertragstitel: title of project.
- Land/Region: country or region, where project takes place at.
- OEZA/ADA-Vertragssumme: amount of money granted by contract.
- Vertragspartner: partner(s) in the project.

### The project pages
When you click on the contract titel in a table you get to the project page. It consists of the same data as the table view, except the additional description text (named "Beschreibung").

### Soundness
So far, we can not say anything about the data quality (completeness, accurateness, etc.), but there are also so far no reaseons to doubt the quality.

**Data errors found**
- [Land/Region missing](http://www.entwicklung.at/zahlen-daten-und-fakten/projektliste/?tx_sysfirecdlist_pi1[showUid]=486&cHash=bcfc60e39b1543897f5492737913c8f0) 
- [Land/Region and Vertragsnummer missing](http://www.entwicklung.at/zahlen-daten-und-fakten/projektliste/?tx_sysfirecdlist_pi1%5BshowUid%5D=1249&cHash=0b453e3503d1f7f1e2e852ea2dece833)
- [Vertragsnummer missing](http://www.entwicklung.at/zahlen-daten-und-fakten/projektliste/?tx_sysfirecdlist_pi1%5BshowUid%5D=942&cHash=9493e43644fc91b324711888c3ea54c2)
- [Vertragssumme is in Partner field, Partner is missing](http://www.entwicklung.at/zahlen-daten-und-fakten/projektliste/?tx_sysfirecdlist_pi1[showUid]=1092&cHash=865e0122108c1411da8cc5c588441d73)
- [Vertragssumme is missing](http://www.entwicklung.at/zahlen-daten-und-fakten/projektliste/?tx_sysfirecdlist_pi1%5BshowUid%5D=646&cHash=fb5bb6e76b698da959f8894b5a417ce9)

## DATA OUTPUT

**raw html**

The scraper downloads all raw html of each table and each project page.

**aid data JSON**

The parsed data is stored in an easy-to-read JSON file for further usage.
```
[
	{
		'contract-number': contract number of the project
		'contract-title': title of the project
		'country-region': country and/or region, where the project takes place
		'OEZA-ADA-contract-volume': amount of funding by austrian development agency
		'contract-partner': partner organisation(s)
		'description': description text of the project
		'url': url of the project page
	},
]
```

**aid data csv**

The parsed data is stored in a human-readable CSV file for further usage.

columns (see attribute description above):
- contract-number
- contract-title
- OEZA-ADA-contract-volume
- contract-partner
- country-region
- description
- url

row: one project each row.

## CONTRIBUTION
In the spirit of free software, everyone is encouraged to help improve this project.

Here are some ways you can contribute:

- by reporting bugs
- by suggesting new features
- by translating to a new language
- by writing or editing documentation
- by analyzing the data
- by visualizing the data
- by writing code (**no pull request is too small**: fix typos in the user interface, add code comments, clean up inconsistent whitespace)
- by refactoring code
- by closing issues
- by reviewing pull requests
- by enriching the data with other data sources

When you are ready, submit a [pull request](https://github.com/OKFNat/aidScraper/pulls).

### Submitting an Issue

We use the [GitHub issue tracker](https://github.com/OKFNat/aidScraper/issues) to track bugs and features. Before submitting a bug report or feature request, check to make sure it hasn't already been submitted. When submitting a bug report, please try to provide a screenshot that demonstrates the problem. 

## COPYRIGHT
All content is openly licensed under the [Creative Commons Attribution 4.0](http://creativecommons.org/licenses/by/4.0/) license, unless otherwisely stated.

All sourcecode is free software: you can redistribute it and/or modify it under the terms of the MIT License.

This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

Visit [http://opensource.org/licenses/MIT](http://opensource.org/licenses/MIT) to learn more about the MIT License.

## REPOSITORY
- [README.md](README.md): Overview of repository
- [code/aid-scraper.py](code/aid-scraper.py): scraper script in python
- [CHANGELOG.md](README.md): Overview of repository
- [LICENSE](README.md): MIT license text

## TODO
**important**
- verify the data
- research: is there a difference between approved funding and paid one?
- convert code to Python3: pay attention to encoding issues in i/o operations

**optional**
- create dataset for network analyses: json, csv for gephi and networkX
- compare data from tables with data from project pages.
- solve the encoding and CR issues. When saving the html table cells to the dict, it is done as unicode, so i did not work out how to replace the '\n' and '\r' characters. Did it then before storing to the CSV file, which is a quick and dirty solution. 

**new features**
- analyze and visualize the data: networkX, maps, Gephi
- add country namecodes for easier combinating with other data

## ACTUAL VERSION
See the [whole history](CHANGELOG.md).

### Version 0.3 - 2016-04-19
**extended scraper**
- aid-scraper.py: fixed the csv output bug caused by cariage return characters.
- update the README.md: add description of scraper.





CHANGELOG.md
## CHANGELOG
### Version 0.3 - 2016-04-19
**extended scraper**
- aid-scraper.py: fixed the csv output bug caused by cariage return characters.
- update the README.md: add description of scraper.

### Version [0.2](https://github.com/OKFNat/aidScraper/commit/bfd7517f650090ce6d460f10c23a3f5d4fa4207a) - 2016-04-13
**extended scraper**
- complete re-structuring of functionality
- add csv export
- add json export
- add README.md
- add documentation and comments
- add LICENSE file

### Version [0.1](https://github.com/OKFNat/aidScraper/commit/90083246be74583e6365165d4b12b9b753f439b7) - 2015-10-29
**init repo**







Scrape_ADC.py
#!/usr/bin/python
# -*- coding: utf8 -*-

from urllib.request import urlopen
from bs4 import BeautifulSoup
import bs4
import requests
from datetime import datetime
import datetime
import csv
import time
import re
import json

def save_to_file(data, filename):
    text_file = open(filename, "w")
    text_file.write(data)
    text_file.close()
    
def getlinks ():
	addresses = []
	for counter in range (0, 45): #when this scraper was written, there were altogether 44 pages
		print ('getting links page '  + str (counter))
		txt_doc='http://www.entwicklung.at/nc/zahlen-daten-und-fakten/projektliste/?tx_sysfirecdlist_pi1%5Btest%5D=test&tx_sysfirecdlist_pi1%5Bmode%5D=1&tx_sysfirecdlist_pi1%5Bsort%5D=uid%3A1&tx_sysfirecdlist_pi1%5Bpointer%5D=' + str(counter)
		soup = BeautifulSoup(urlopen(txt_doc), 'html.parser')
		for item in soup.find_all ('a'):
			if 'href' in item.attrs:
				if "http://www.entwicklung.at/zahlen-daten-und-fakten/projektliste/?" in item['href']:
					addresses.append (item['href'])
	return addresses

def striphtml (data):
    p = re.compile(r'<.*?>')
    return p.sub('', data)
    

if __name__ == "__main__":
	addresses = getlinks ()
	data = []
	
	l = len (addresses)
	k = 1
	print ("downloaded %s addresses" % str (l))
	
	for address in addresses:
		elem = {}
		response = requests.get (address)
		soup2 = bs4.BeautifulSoup (response.text)
		content = soup2.find_all ('div', {'cdcontentdiv'})
		
		aktualisierung = soup2.find_all ('div', id= 'c3936')
		div = aktualisierung [0].find_all('div')
		date=div[0].contents
		Aktualisierungsdatum = date[0].replace('\xa0', '') 
		
		elem["Link"] = address
		elem["Vertragsnummer"] = striphtml (str (content [0]))
		elem["Vertragstitel"]= striphtml (str (content [1]))
		elem["LandRegion"]= striphtml (str (content [2]))
		elem["Vertragspartner"]= striphtml (str (content [3]))
		elem["Vertragssumme"]= striphtml (str (content [4]))
		elem["Beschreibung"]= striphtml (str(content [5]))
		ts = time.time()
		elem["Timestamp"] = datetime.datetime.fromtimestamp(ts).strftime('%Y-%m-%d %H:%M:%S')
		elem["Aktualisierungsdatum"]=Aktualisierungsdatum
		print ("downloaded %s of %s entries" % (k, l))
		k = k+1
		
		data.append(elem)
	save_to_file(json.dumps(data, indent=2), 'Entwicklunghilfe.json')


code/aid-scraper.py
#!/bin/env python2
# -*- coding: utf-8 -*-

"""
Fetches, parses and stores the data from the austrian aid projects website since January 1st 2010. 
"""

import urllib2
import bs4
from bs4 import BeautifulSoup
from datetime import datetime
import re
import json
import os
import string
import time

__author__ = "Christian Goebel, Stefan Kasberger"
__copyright__ = "Copyright 2015"
__license__ = "MIT"
__version__ = "0.3"
__maintainer__ = "Stefan Kasberger"
__email__ = "mail@stefankasberger.at"
__status__ = "Production" # 'Development', 'Production' or 'Prototype'

###    GLOBAL   ###

ROOT_FOLDER = os.path.dirname(os.getcwd())
FOLDER_RAW_HTML = ROOT_FOLDER + '/data/raw/html/'
FOLDER_CSV = ROOT_FOLDER + '/data/csv/'
FOLDER_JSON = ROOT_FOLDER + '/data/json/'
FILENAME_BASE = 'aid-data'
BASE_URL = 'http://www.entwicklung.at/nc/zahlen-daten-und-fakten/projektliste/'
QUERY_URL = BASE_URL + '?tx_sysfirecdlist_pi1[test]=test&tx_sysfirecdlist_pi1[mode]=1&tx_sysfirecdlist_pi1[sort]=uid%3A1&tx_sysfirecdlist_pi1[pointer]='
DELAY_TIME = 5 # in seconds
# TS = datetime.now().strftime('%Y-%m-%d-%H-%M')
TS = '2016-04-03-03-05'

###    FUNCTIONS   ###

def SetupEnvironment():
	"""Sets up the folder structure and working environment.
	"""
	if not os.path.exists(FOLDER_RAW_HTML):
		os.makedirs(FOLDER_RAW_HTML)
	if not os.path.exists(FOLDER_JSON):
		os.makedirs(FOLDER_JSON)
	if not os.path.exists(FOLDER_CSV):
		os.makedirs(FOLDER_CSV)

def FetchHtml(url):
	"""Fetches html url via urllib2.
	
	Args:
		url: url to fetch (string).
	
	Returns:
		html: html string as unicode.
	"""
	response = urllib2.urlopen(url)
	html = response.read().decode('utf-8')
	time.sleep(DELAY_TIME)
	
	return html

def FetchHtmlTables(url, folder, pageCounter = 0):
	"""Fetches each overview page with the table of the projects and saves the html locally.
	
	Args:
		url: string to fetch.
		folder: directory to save the html files in.
		pageCounter: counter for the actual number of the page (int).
	"""
	# the url has a page-counter at the end, so just increase the counter and fetch one page after another
	rawHtml = FetchHtml(url+str(pageCounter))

	if not os.path.exists(folder+TS):
		os.makedirs(folder+TS)
	Save2File(rawHtml, folder+TS+'/'+TS+'_table-'+str(pageCounter)+'.html')
	# checks if there is a "weiter" (more) anchor, cause this means a next page exists
	isWeiter = re.findall(r'<a href="(.*)" >weiter &gt;</a></div></div>', rawHtml)
	if isWeiter:
		pageCounter += 1
		FetchHtmlTables(url, folder, pageCounter)

def FetchHtmlProjects(aidData, folder):
	"""Fetches html of the project pages and saves the html locally.
	
	Args:
		aidData: list[] of dict{} of aid projects.
		folder: directory to save the html files in.
	
	Returns:
		aidData: list[] of dict{} of aid projects.
	"""
	if not os.path.exists(folder+TS):
		os.makedirs(folder+TS)
	# iterate over every list element to fetch and store the project page
	for idx, project in enumerate(aidData):
		html = FetchHtml(project['url'])
		Save2File(html, folder+TS+'/'+TS+'_project-'+str(idx)+'.html')

	return aidData

def Save2File(data, filename):
	"""Saves a file locally.
	
	Args:
		data: string to save.
		filename: name of the file with path.
	"""
	text_file = open(filename, "w")
	text_file.write(data.encode('utf-8'))
	text_file.close()

def ReadFile(filename):
	"""Reads a file and returns the string.
	
	Args:
		filename: name of the file
	
	Returns:
		string: file as text-string.
	"""
	f = open(filename, 'r')
	string = f.read()
	
	return string

def ReadTableFilesInFolder(folder):
	"""Reads in all table html-files from a folder.
	
	Args:
		folder: folder where the html-files are stored in.
	
	Returns:
		sortedList: list[] of sorted html texts.
	"""
	htmlList = []
	sortedList = []

	for filename in os.listdir(folder):
		if filename.find(TS+'_table') >= 0:
			rawHtml = ReadFile(folder+'/'+filename)
			fileIndex = filename.split('.')[0].split('-')[5]
			htmlList.append((int(fileIndex), rawHtml))
	# sort list of duppels after first element (the filename postfix) and store to list[]
	htmlList = sorted(htmlList, key=lambda htmlList: htmlList[0])
	for idx, html in htmlList:
		sortedList.append(html)

	return sortedList

def ReadProjectFilesInFolder(aidData, folder):
	"""Reads in all table html-files from a folder.
	
	Args:
		aidData: list[] of dict{} of aid projects.
		folder: folder where the html-files are stored in.
	
	Returns:
		sortedList: list[] of sorted html texts.
	"""
	htmlList = []
	sortedList = []

	for filename in os.listdir(folder):
		if filename.find(TS+'_project') >= 0:
			rawHtml = ReadFile(folder+'/'+filename)
			fileIndex = filename.split('.')[0].split('-')[5]
			htmlList.append((int(fileIndex), rawHtml))
	
	# sort list of duppels after first element (the filename postfix) and store to list[]
	htmlList = sorted(htmlList, key=lambda htmlList: htmlList[0])
	for idx, html in htmlList:
		sortedList.append(html)

	return sortedList

def ParseTables(html):
	"""Parses each row of the html table.
	
	Args:
		html: list[] of duppels (filename, html).
	
	Returns:
		aidData: list[] of dict{} of aid projects.
			'contract-number': contract number.
			'contract-title': title of the project.
			'country-region': country and/or region, where the project takes place.
			'OEZA-ADA-contract-volume': amount of funding.
			'contract-partner': partner organisation(s).
			'url': url of the project page.
	"""
	aidData = []
	counter = 1

	for page in html:
		soup = BeautifulSoup(page)
		table = soup.find_all('table')[1]
		for row in table.find_all('tr')[1:]:
			project = {}
			tds = row.find_all('td')
			project['unique-id'] = str(counter)
			project['contract-title'] = tds[1].div.text
			project['url'] =tds[1].div.a['href']
			project['country-region'] = tds[2].div.text
			project['OEZA-ADA-contract-volume'] = tds[3].div.text
			project['contract-partner'] = tds[4].div.text
			project['contract-number'] = tds[0].div.text
			aidData.append(project)
			counter += 1
	
	return aidData

def ParseProjects(aidData, htmlProjects):
	"""Parses the description of every project page.
	
	Args:
		aidData: list[] of dict{} of aid projects.
		htmlProjects: list[] of html texts.
	
	Returns:
		aidData: list[] of dict{} of aid projects.
			'unique-id': unique id for each project.
			'contract-number': contract number.
			'contract-title': title of the project.
			'country-region': country and/or region, where the project takes place.
			'OEZA-ADA-contract-volume': amount of funding.
			'contract-partner': partner organisation(s).
			'description': description text of the project.
			'url': url of the project page.
	"""
	for idx, html in enumerate(htmlProjects):
		soup = BeautifulSoup(html)
		aidData[int(idx)]['description'] = soup.find_all('table')[0].find_all('tr')[5].find_all('div')[1].text
	
	return aidData

def SaveAidData(aidData, filename):
	"""Saves the aid data as JSON file.
	
	Args:
		aidData: list[] of dict{} of aid projects.
		filename: filepath of the JSON file. 
	"""
	Save2File(json.dumps(aidData, indent=2, ensure_ascii=False), filename)
	print 'Aid data exported as JSON:',filename

def OpenAidData(filename):
	"""Opens the aid data JSON file.
	
	Args:
		filename: filepath of the JSON file.
	"""
	aidData = json.loads(ReadFile(filename))
	
	return aidData

def Save2CSV(data, filename):
	"""Exports the aid data into a structured csv file.
	
	Args:
		data: list[] of dict{} of aid projects.
		filename: filepath.
	"""
	csvString = '"unique-id","contract-number","contract-title","OEZA-ADA-contract-volume","contract-partner","country-region","description","url"\n'

	# iterate over each project
	for project in data:
		uniqueId = '""'
		number = '""'
		title = '""'
		funding = '""'
		partner = '""'
		region = '""'
		description = '""'
		url = '""'
		
		# read out each attribute
		for elem in project.keys():
			val = project[elem].replace('"', '').replace('\n', '').replace('\r', '') # replace apostrophes.
			if elem == 'unique-id':
				uniqueId = '"'+val+'"'
			if elem == 'contract-number':
				number = '"'+val+'"'
			if elem == 'contract-title':
				title = '"'+val+'"'
			if elem == 'OEZA-ADA-contract-volume':
				funding = '"'+val+'"'
			if elem == 'contract-partner':
				partner = '"'+val+'"'
			if elem == 'country-region':
				region = '"'+val+'"'
			if elem == 'description':
				description = '"'+val.strip('\n\r')+'"'
			if elem == 'url':
				url = '"'+val+'",'
			
		csvString += uniqueId+','+number+','+title+','+funding+','+partner+','+region+','+description+','+url+'\n'

	Save2File(csvString, filename)
	print 'Aid data exported as CSV:',filename

if __name__ == "__main__":

	# setup
	startTime = datetime.now()
	print 'start:', startTime
	aidData = []
	SetupEnvironment()
	DOWNLOAD_FILES = False
	PARSE_FILES = False
	STRUCTURE_DATA = False

	if DOWNLOAD_FILES:
		FetchHtmlTables(QUERY_URL, FOLDER_RAW_HTML) # html as string
		htmlTables = ReadTableFilesInFolder(FOLDER_RAW_HTML+TS) # html as string
		aidData = ParseTables(htmlTables)
		SaveAidData(aidData, FOLDER_JSON+'aid-data_'+TS+'.json')
		FetchHtmlProjects(aidData, FOLDER_RAW_HTML)
	
	if PARSE_FILES:
		htmlTables = ReadTableFilesInFolder(FOLDER_RAW_HTML+TS) # html as string
		aidData = ParseTables(htmlTables)
		htmlProjects = ReadProjectFilesInFolder(aidData, FOLDER_RAW_HTML+TS) # html as string
		aidData = ParseProjects(aidData, htmlProjects)
		SaveAidData(aidData, FOLDER_JSON+'aid-data_'+TS+'.json')

	if STRUCTURE_DATA:
		aidData = OpenAidData(FOLDER_JSON+'aid-data_'+TS+'.json')
		Save2CSV(aidData, FOLDER_CSV+'aid-data_'+TS+'.csv')
	
	print 'runtime:', (datetime.now() - startTime)
